crates/faucet/build.rs

```diff
@@ -0,0 +1,92 @@
+use std::env;
+use std::fs::{self};
+use std::io::{self};
+use std::path::{Path, PathBuf};
+
+use miden_client::transaction::{TransactionKernel, TransactionScript};
+use miden_client::utils::Serializable;
+
+const ASSETS_DIR: &str = "assets";
+const ASM_DIR: &str = "asm";
+const ASM_TX_SCRIPTS_DIR: &str = "tx_scripts";
+
+/// Compile contents of asm directory into .txs files.
+fn main() {
+    // re-build when the MASM code changes
+    println!("cargo::rerun-if-changed={ASM_DIR}/");
+
+    let crate_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
+    let source_dir = Path::new(&crate_dir).join(ASM_DIR);
+
+    let build_dir = env::var("OUT_DIR").unwrap();
+    let target_dir = Path::new(&build_dir).join(ASSETS_DIR);
+
+    compile_transaction_scripts(
+        &source_dir.join(ASM_TX_SCRIPTS_DIR),
+        &target_dir.join(ASM_TX_SCRIPTS_DIR),
+    );
+}
+
+/// Reads all MASM files from the `source_dir`, compiles each file individually into a TXS
+/// file, and stores the compiled files into the `target_dir`.
+///
+/// The source files are expected to contain executable programs.
+fn compile_transaction_scripts(source_dir: &Path, target_dir: &Path) {
+    fs::create_dir_all(target_dir).expect("should create target directory");
+    let assembler = TransactionKernel::assembler();
+
+    let masm_files = get_masm_files(source_dir).expect("should find MASM files");
+    for masm_file_path in masm_files {
+        // read the MASM file, parse it, and serialize the parsed AST to bytes
+        let code = assembler
+            .clone()
+            .assemble_program(masm_file_path.clone())
+            .expect("program should assemble correctly");
+        let script = TransactionScript::new(code);
+
+        let bytes = script.to_bytes();
+
+        let masm_file_name = masm_file_path.file_name().expect("file name should exist");
+        let mut txs_file_path = target_dir.join(masm_file_name);
+
+        // write the binary TXS to the output dir
+        txs_file_path.set_extension("txs");
+        fs::write(txs_file_path, bytes).expect("should write .txs file");
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Returns a vector with paths to all MASM files in the specified directory.
+///
+/// All non-MASM files are skipped.
+fn get_masm_files(dir_path: &Path) -> io::Result<Vec<PathBuf>> {
+    let mut files = Vec::new();
+
+    let entries = fs::read_dir(dir_path)?;
+    for entry in entries {
+        let file_path = entry?.path();
+        if is_masm_file(&file_path)? {
+            files.push(file_path);
+        }
+    }
+
+    Ok(files)
+}
```

Comment:
Perhaps a broader point (not for this PR) is whether we could re-use that code from miden-lib. Even there we have two places where `get_masm_files` is defined: in `build.rs` and in `src/transaction/mod.rs`.
Sadly we can't re-use the code easily because `build.rs` can't depend on code defined in its own (yet un-built) crate.
An interim solution would be to extract the definition from `src/transaction/mod.rs` to say, `src/utils/', and we could reuse it both in `src/transaction/mod.rs` and here (and `build.rs` keeps its own definition).
Another solution would be to introduce a helper crate where all three locations could pull `get_masm_files` from - but that's an overkill unless we have other uses for this helper crate (and also it's not too much code that is duplicated).

Reply:
Agreed that it would be good to encapsulate this logic a bit better - but I'm viewing this as a temporary solution on our way to figure out out how to structure such projects properly. The longer-term solution would be to put the script itself into a separate crate (and wrap it into a "package") to align the methodology with how the compiler will generate package.